# Part 1. Initial EDA and data cleaning

In [4]:
# Import libraries and shared functions
from helpers import *

In [5]:
# Load dataset into a dataframe
df_raw = pd.read_csv("HR_comma_sep.csv")

# Display first few rows of the dataframe
df_raw.head(10)

satisfaction_level  last_evaluation  number_project  average_montly_hours  \
0                0.38             0.53               2                   157   
1                0.80             0.86               5                   262   
2                0.11             0.88               7                   272   
3                0.72             0.87               5                   223   
4                0.37             0.52               2                   159   
5                0.41             0.50               2                   153   
6                0.10             0.77               6                   247   
7                0.92             0.85               5                   259   
8                0.89             1.00               5                   224   
9                0.42             0.53               2                   142   

   time_spend_company  Work_accident  left  promotion_last_5years Department  \
0                   3              0     1                      0      sales   
1                   6              0     1                      0      sales   
2                   4              0     1                      0      sales   
3                   5              0     1                      0      sales   
4                   3              0     1                      0      sales   
5                   3              0     1                      0      sales   
6                   4              0     1                      0      sales   
7                   5              0     1                      0      sales   
8                   5              0     1                      0      sales   
9                   3              0     1                      0      sales   

   salary  
0     low  
1  medium  
2  medium  
3     low  
4     low  
5     low  
6     low  
7     low  
8     low  
9     low

In [6]:
# Gather basic information about the data
df_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 10 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   satisfaction_level     14999 non-null  float64
 1   last_evaluation        14999 non-null  float64
 2   number_project         14999 non-null  int64  
 3   average_montly_hours   14999 non-null  int64  
 4   time_spend_company     14999 non-null  int64  
 5   Work_accident          14999 non-null  int64  
 6   left                   14999 non-null  int64  
 7   promotion_last_5years  14999 non-null  int64  
 8   Department             14999 non-null  object 
 9   salary                 14999 non-null  object 
dtypes: float64(2), int64(6), object(2)
memory usage: 1.1+ MB


In [7]:
# Gather descriptive statistics about the data
df_raw.describe()

satisfaction_level  last_evaluation  number_project  \
count        14999.000000     14999.000000    14999.000000   
mean             0.612834         0.716102        3.803054   
std              0.248631         0.171169        1.232592   
min              0.090000         0.360000        2.000000   
25%              0.440000         0.560000        3.000000   
50%              0.640000         0.720000        4.000000   
75%              0.820000         0.870000        5.000000   
max              1.000000         1.000000        7.000000   

       average_montly_hours  time_spend_company  Work_accident          left  \
count          14999.000000        14999.000000   14999.000000  14999.000000   
mean             201.050337            3.498233       0.144610      0.238083   
std               49.943099            1.460136       0.351719      0.425924   
min               96.000000            2.000000       0.000000      0.000000   
25%              156.000000            3.000000       0.000000      0.000000   
50%              200.000000            3.000000       0.000000      0.000000   
75%              245.000000            4.000000       0.000000      0.000000   
max              310.000000           10.000000       1.000000      1.000000   

       promotion_last_5years  
count           14999.000000  
mean                0.021268  
std                 0.144281  
min                 0.000000  
25%                 0.000000  
50%                 0.000000  
75%                 0.000000  
max                 1.000000

In [8]:
# Display all column names
df_raw.columns

Index(['satisfaction_level', 'last_evaluation', 'number_project',
       'average_montly_hours', 'time_spend_company', 'Work_accident', 'left',
       'promotion_last_5years', 'Department', 'salary'],
      dtype='object')

In [9]:
# Standardize column names in snake_case
# Rename columns
df_raw.rename(columns={ "Work_accident" : "work_accident",
                    "Department": "department",
                    "average_montly_hours": "average_monthly_hours",
                    "time_spend_company": "tenure",
                    "promotion_last_5years" : "promotion_last_5_years"
                     
}, inplace=True)

# Display all column names after the update
df_raw.columns

Index(['satisfaction_level', 'last_evaluation', 'number_project',
       'average_monthly_hours', 'tenure', 'work_accident', 'left',
       'promotion_last_5_years', 'department', 'salary'],
      dtype='object')

In [10]:
# Check for missing values
df_raw.isna().sum()

satisfaction_level        0
last_evaluation           0
number_project            0
average_monthly_hours     0
tenure                    0
work_accident             0
left                      0
promotion_last_5_years    0
department                0
salary                    0
dtype: int64

In [11]:
# Check for duplicates
df_raw.duplicated().sum()

3008

In [12]:
# Inspect some rows containing duplicates
df_raw[df_raw.duplicated()].head(10)

satisfaction_level  last_evaluation  number_project  \
396                  0.46             0.57               2   
866                  0.41             0.46               2   
1317                 0.37             0.51               2   
1368                 0.41             0.52               2   
1461                 0.42             0.53               2   
1516                 0.40             0.50               2   
1616                 0.37             0.46               2   
1696                 0.39             0.56               2   
1833                 0.10             0.85               6   
12000                0.38             0.53               2   

       average_monthly_hours  tenure  work_accident  left  \
396                      139       3              0     1   
866                      128       3              0     1   
1317                     127       3              0     1   
1368                     132       3              0     1   
1461                     142       3              0     1   
1516                     127       3              0     1   
1616                     156       3              0     1   
1696                     160       3              0     1   
1833                     266       4              0     1   
12000                    157       3              0     1   

       promotion_last_5_years  department  salary  
396                         0       sales     low  
866                         0  accounting     low  
1317                        0       sales  medium  
1368                        0       RandD     low  
1461                        0       sales     low  
1516                        0          IT     low  
1616                        0       sales     low  
1696                        0       sales     low  
1833                        0       sales     low  
12000                       0       sales     low

It is unlikely that the observations are having the same values with several continuous variables across 10 columns. 
Most likely these records are not legitimate, thus proceeding to drop them.

In [13]:
# Drop duplicates
df = df_raw.drop_duplicates()

# Display first rows of new dataframe
df.head(10)

satisfaction_level  last_evaluation  number_project  average_monthly_hours  \
0                0.38             0.53               2                    157   
1                0.80             0.86               5                    262   
2                0.11             0.88               7                    272   
3                0.72             0.87               5                    223   
4                0.37             0.52               2                    159   
5                0.41             0.50               2                    153   
6                0.10             0.77               6                    247   
7                0.92             0.85               5                    259   
8                0.89             1.00               5                    224   
9                0.42             0.53               2                    142   

   tenure  work_accident  left  promotion_last_5_years department  salary  
0       3              0     1                       0      sales     low  
1       6              0     1                       0      sales  medium  
2       4              0     1                       0      sales  medium  
3       5              0     1                       0      sales     low  
4       3              0     1                       0      sales     low  
5       3              0     1                       0      sales     low  
6       4              0     1                       0      sales     low  
7       5              0     1                       0      sales     low  
8       5              0     1                       0      sales     low  
9       3              0     1                       0      sales     low

In [14]:
# Look for outliers in 'tenure' column

# Compute the 25th percentile value in 'tenure'
q1 = df['tenure'].quantile(0.25)

# Compute the 75th percentile value in 'tenure'
q3 = df['tenure'].quantile(0.75)

# Calculate the interquartile range for 'tenure' column
iqr = q3 - q1

# Define outliers as data points that fall below Q1 - 1.5*IQR or above Q3 + 1.5*IQR
upper_limit = q3 + 1.5 * iqr
lower_limit = q1 - 1.5 * iqr
print("Lower limit:", lower_limit)
print("Upper limit:", upper_limit)

Lower limit: 1.5
Upper limit: 5.5


In [15]:
# Determine the number of rows containing outliers
outliers = df[(df['tenure'] > upper_limit) | (df['tenure'] < lower_limit)]
outliers.shape[0]

824

Number of rows containing outliers in 'tenure' column: 824.

Will consider it when constructing models that are sensitive to outliers.

In [16]:
# Save data frame after initial cleaning
df.to_csv('hr_data.csv', index=False)